### PR TITLE
Keep `Labels.values` on the initial device in TorchScript

### LINKED ISSUE #290: Bug or undocumented behavior: Creation of TensorBlock moves Labels device to cpu
Hope it is clear from title, here code snippet

```python
import torch
from equistore.torch import Labels, TensorBlock

sample_labels = Labels(["s"], torch.tensor([[0], [2], [4]], device='cuda:0'))
print(sample_labels.values.device) # cuda:0
block_1 = TensorBlock(
    values=torch.full((3, 1, 1), 1.0, device='cuda:0'),
    samples=sample_labels,
    components=[Labels(["c"], torch.tensor([[0]], device='cuda:0'))],
    properties=Labels(["p"], torch.tensor([[0]], device='cuda:0')),
)
print(block_1.samples.values.device) # cpu
```

### PR DESCRIPTION
**Pull request by Luthaf**
Fix #290 

----
:books: Documentation preview :books:: https://equistore--291.org.readthedocs.build/en/291/

### CODE REVIEW

File: equistore/src/labels.rs [hunk 1]
```diff
@@ -135,8 +135,8 @@ pub struct Labels {
 // Labels can be sent to other thread safely since eqs_labels_t uses an
 // `Arc<equistore_core::Labels>`, so freeing them from another thread is fine
 unsafe impl Send for Labels {}
-// &Labels can be sent to other thread safely since there is no interior
-// mutability
+// &Labels can be sent to other thread safely since there is no un-synchronized
+// interior mutability (`user_data` is protected by RwLock).
```
[the remark this answers is not in the document]
Comment (Luthaf) [reply equistore/src/labels.rs:139]: As mentioned in the commit message, these functions are not yet exposed in the Rust/Python bindings, since I don't see a use case for them. We can always add them later if we need them!

File: equistore-torch/src/labels.cpp [hunk 2]
```diff
@@ -126,27 +126,44 @@ static std::vector<std::string> names_from_equistore(const equistore::Labels& la
...
         sizes,
         torch::TensorOptions().dtype(torch::kInt32)
     );
 }
 
 LabelsHolder::LabelsHolder(torch::IValue names, torch::Tensor values):
     names_(details::normalize_names(names, "names")),
-    values_(normalize_int32_tensor(std::move(values), 2, "Labels values"))
+    values_(normalize_int32_tensor(std::move(values), 2, "Labels values")),
+    labels_(equistore::Labels(labels_from_torch(names_, values_)))
```
[the remark this answers is not in the document]
Comment (Luthaf) [reply equistore-torch/src/labels.cpp:158]: yes, good catch! I tried multiple things, and got mixed up

File: equistore-core/src/c_api/labels.rs [hunk 3]
```diff
@@ -213,6 +213,76 @@ pub unsafe extern fn eqs_labels_create(
...
+/// If no data has been registered, `*user_data` will be NULL.
+///
+/// @param labels set of labels containing user data
+/// @param user_data this will be set to the pointer than was registered with
+///                  these `labels`
+/// @returns The status code of this operation. If the status is not
+///          `EQS_SUCCESS`, you can use `eqs_last_error()` to get the full
+///          error message.
+#[no_mangle]
+pub unsafe extern fn eqs_labels_get_user_data(
```
[the remark this answers is not in the document]
Comment (Luthaf) [reply equistore-core/src/c_api/labels.rs:265]: good point, I'll rename this function to `eqs_labels_user_data`

**Review by agoscinski — COMMENTED** [review]
Tested branch on my machine, works.

I am a bit worried of mismatches between the CPU copy and the GPU values due to inplace operations. I guess these operations cause a mismatch between the CPU and GPU version?
```python
samples.values[:] = 1
```

```python
torch.square(samples.values, out=samples.values)
```

If I am correct, I am not sure how to make the user aware of this or handle this in a smart way. Do you have any idea?

**Comment by Luthaf** [comment]
So sample values should be treated as immutable (and they are in the Python version). But until PyTorch gives us a way to mark tensor as immutable (https://github.com/pytorch/pytorch/issues/44027) there is not much we can do. I'll add a note in the documentation about this.

**Comment by Luthaf** [comment]
Here is the warning about immutability: https://github.com/lab-cosmo/equistore/pull/291/files#diff-aa492194763b217c46428ff2db48416ee9e4cf472bf610bc65a995d2a14cf7e9R47-R57
